### ISSUE TITLE
Full std::type_index interface support

### ISSUE THREAD

**Issue by foonathan** [Manu343726/ctti#3]
`ctti::type_index` should have the full interface as `std::type_index`. Currently only missing the comparison operators.
Also, `name()` should really return a `const char*` but this has the null-termination issue discussed in my PR.

**Comment by Manu343726** [comment]
I'm agree with this. About `name()` returning `ctti::detail::string`, that was a try in the line of "`ctti::detail::string` only", the first alternative I were talking about in your PR.

**Comment by Manu343726** [comment]
I have been checking docs from both `std::type_index` and `ctti::type_info`. Since ctti features have no runtime overhead, there's no point for me in wrapping a `type_info` to get a `type_index`, I would just provide features of both at once (What `ctti::type_id_t` currently does). 

That said, `type_id_t` misses the `before()` function from `type_info`, plus comparison operators. Since the collation order is implementation defined, it offers no order guarantees. I was thinking on using hash for comparison in addition to in/equality. Except from MSVC putting "struct" in front of `struct` type names, our names should be equal on all platforms, yielding same hash, guaranteeing same type ordering.

**Comment by foonathan** [comment]
Couldn't we remove the prefix to have the same name? Otherwise I agree.

**Comment by Manu343726** [comment]
Of course we can, but it may be expensive: 

``` cpp
if(name.substr<0,8>() == "struct ")
    return name.substr<8,name.size()>();
else
    return name;
```

**Comment by foonathan** [comment]
Then pack an `#ifdef _MSC_VER` around it :D

Maybe you can avoid the branch then completely, is the `struct` before each type?

**Comment by Manu343726** [comment]
No, for classes it prepends with `class`, and no prefix is added for basic types. If `struct` and `class` had the same length we could use a trait to check for prefix. Also is not that simple since when the type has cv qualifiers `struct`/`class` is no longer a prefix.

**Comment by foonathan** [comment]
Damn, then at least make the check only on MSVC. A hash that doesn't change with the platform would be great.
